scio-core/src/main/scala/com/spotify/scio/util/BloomFilter.scala

```diff
@@ -0,0 +1,576 @@
+      //
+      // However, the true set cardinality may not be known. From empirical evidence, though,
+      // it is upper bounded with high probability by 1.1 * estimatedCardinality (as long as the
+      // Bloom filter is not too full), so we plug this into the formula instead.
+      // TODO: investigate this upper bound and density more closely (or derive a better formula).
+      // TODO: The following logic is same for immutable Bloom Filters and may be referred here.
+      val fpProb =
+        if (density > 0.95)
+          1.0 // No confidence in the upper bound on cardinality.
+        else
+          scala.math.pow(1 - scala.math.exp(-numHashes * size.estimate * 1.1 / width), numHashes)
+
+      ApproximateBoolean(true, 1 - fpProb)
+    } else {
+      // False negatives are not possible.
+      ApproximateBoolean.exactFalse
+    }
+
+  /**
+   * This may be faster if you don't care about evaluating
+   * the false positive probability
+   */
+  def maybeContains(item: A): Boolean
+
+  // Estimates the cardinality of the set of elements that have been
+  // inserted into the Bloom Filter.
+  def size: Approximate[Long]
+
+  def toBitSet: util.BitSet
+
+  def copy: MutableBF[A]
+
+  /**
+   * Compute the Hamming distance between the two Bloom filters
+   * `a` and `b`. The distance is defined as the number of bits that
+   * need to change to in order to transform one filter into the other.
+   * This is computed using XOR but it doesn't mutate any BloomFilters
+   */
+  def hammingDistance(that: MutableBF[A]): Int =
+    (this, that) match {
+      // Comparing with empty filter should give number
+      // of bits in other set
+      case (x: MutableBFZero[A], y: MutableBFZero[A]) => 0
+      case (x: MutableBFZero[A], y: MutableBF[A])     => y.numBits
+      case (x: MutableBF[A], y: MutableBFZero[A])     => x.numBits
+
+      // Otherwise compare as bit sets
+      case (_, _) =>
+        // hammingDistance should not mutate BloomFilter
+        val thisCopy = this.toBitSet.clone().asInstanceOf[util.BitSet]
+        thisCopy.xor(that.toBitSet)
+        thisCopy.cardinality()
+    }
+
+}
+
+/**
+ * Empty bloom filter.
+ */
+private[scio] final case class MutableBFZero[A](hashes: KirMit32Hash[A]) extends MutableBF[A] {
+
+  lazy val toBitSet: util.BitSet = new util.BitSet()
+
+  lazy val numHashes: Int = hashes.numHashes
+
+  lazy val width: Int = hashes.width
+
+  lazy val numBits: Int = 0
+
+  // scalastyle:off method.name
+  def ++=(other: MutableBF[A]): MutableBF[A] = other
+
+  def +=(other: A): MutableBF[A] = MutableBFInstance[A](hashes, other)
+  // scalastyle:on method.name
+
+  def checkAndAdd(other: A): (MutableBF[A], ApproximateBoolean) =
+    (this += other, ApproximateBoolean.exactFalse)
+
+  override def contains(item: A): ApproximateBoolean = ApproximateBoolean.exactFalse
+
+  def maybeContains(item: A): Boolean = false
+
+  lazy val size: Approximate[Long] = Approximate.exact[Long](0)
```

Comment:
does this need to be lazy?

Reply:
it doesn't, I was trying to some tests on memory allocations, but it wasn't fruitful. I'll revert these.